File: togo/internal/services/semantic_tokens.go
package services

// Semantic tokens: classify every resolved identifier so editors color fields,
// locals, parameters, type parameters etc. accurately. Offset-based and
// position-free; the LSP server encodes the entries into the wire format.
// Port of src/services/semanticTokens.ts.

import (
	"cmp"
	"slices"

	"github.com/nikeee/cappu/internal/compiler"
)

// TokenTypes is the semantic-token legend the server advertises.
var TokenTypes = []string{
	"namespace", "class", "interface", "enum", "enumMember", "typeParameter",
	"type", "method", "property", "parameter", "variable", "regexp",
}

// regexSinks lists standard-library methods whose named string parameters are
// regular expressions, keyed by "OwnerSimpleName#method" -> regex arg indices.
// A string literal passed at one of these positions is tokenized as `regexp`
// so clients can highlight the pattern (Java has no regex literal syntax).
var regexSinks = map[string]map[int]bool{
	"Pattern#compile":     {0: true},
	"Pattern#matches":     {0: true},
	"String#matches":      {0: true},
	"String#split":        {0: true},
	"String#replaceAll":   {0: true},
	"String#replaceFirst": {0: true},
}

// enclosingTypeName returns the simple name of the type declaration enclosing a
// declaration node, or "" if there is none.
func enclosingTypeName(node *compiler.Node) string {
	for n := node.Parent; n != nil; n = n.Parent {
		switch n.Kind {
		case compiler.ClassDeclaration, compiler.InterfaceDeclaration, compiler.EnumDeclaration,
			compiler.RecordDeclaration, compiler.AnnotationTypeDeclaration:
			if name := declName(n); name != nil {
				return name.AsIdentifier().Text
			}
			return ""
		}
	}
	return ""
}

// TokenModifiers is the semantic-token modifier legend.
var TokenModifiers = []string{"declaration", "static", "readonly", "defaultLibrary", "deprecated"}

func typeIndex(name string) int {
	for i, t := range TokenTypes {
		if t == name {
			return i
		}
	}
	return -1
}

const (
	modDeclaration    = 1 << 0
	modStatic         = 1 << 1
	modReadonly       = 1 << 2
	modDefaultLibrary = 1 << 3
	modDeprecated     = 1 << 4
)

// SemanticTokenEntry is one classified identifier span.
type SemanticTokenEntry struct {
	Offset         int
	Length         int
	TokenType      int // index into TokenTypes
	TokenModifiers int // bit set over TokenModifiers
}

func tokenTypeOf(flags compiler.SymbolFlags) int {
	switch {
	case flags&compiler.SymbolFlagsPackage != 0:
		return typeIndex("namespace")
	case flags&compiler.SymbolFlagsClass != 0:
		return typeIndex("class")
	case flags&compiler.SymbolFlagsInterface != 0:
		return typeIndex("interface")
	case flags&compiler.SymbolFlagsEnum != 0:
		return typeIndex("enum")
	case flags&compiler.SymbolFlagsRecord != 0:
		return typeIndex("class")
	case flags&compiler.SymbolFlagsAnnotation != 0:
		return typeIndex("type")
	case flags&compiler.SymbolFlagsTypeParameter != 0:
		return typeIndex("typeParameter")
	case flags&compiler.SymbolFlagsEnumConstant != 0:
		return typeIndex("enumMember")
	case flags&(compiler.SymbolFlagsMethod|compiler.SymbolFlagsConstructor) != 0:
		return typeIndex("method")
	case flags&compiler.SymbolFlagsField != 0:
		return typeIndex("property")
	case flags&compiler.SymbolFlagsParameter != 0:
		return typeIndex("parameter")
	case flags&compiler.SymbolFlagsLocalVariable != 0:
		return typeIndex("variable")
	default:
		return -1
	}
}

// modifierCarrier returns the node whose modifiers govern this symbol.
func modifierCarrier(symbol *compiler.Symbol) *compiler.Node {
	declaration := symbolDeclaration(symbol)
	if declaration == nil {
		return nil
	}
	if declaration.Kind == compiler.VariableDeclarator {
		return declaration.Parent
	}
	return declaration
}

func modifiersOf(symbol *compiler.Symbol, isDeclarationName bool) int {
	bits := 0
	if isDeclarationName {
		bits = modDeclaration
	}
	if symbol.Flags&compiler.SymbolFlagsEnumConstant != 0 {
		bits |= modStatic | modReadonly
	} else if carrier := modifierCarrier(symbol); carrier != nil {
		if mods := declModifiers(carrier); mods != nil {
			for _, m := range mods.Nodes {
				if m.Kind == compiler.StaticKeyword {
					bits |= modStatic
				}
				if m.Kind == compiler.FinalKeyword {
					bits |= modReadonly
				}
			}
		}
	}
	if declaration := symbolDeclaration(symbol); declaration != nil {
		if compiler.IsSyntheticURI(compiler.URI(compiler.GetSourceFileOfNode(declaration).AsSourceFile().FileName)) {
			bits |= modDefaultLibrary
		}
	}
	if _, ok := compiler.SymbolDeprecation(symbol); ok {
		bits |= modDeprecated
	}
	return bits
}

// GetSemanticTokens classifies every resolved identifier in a source file.
func GetSemanticTokens(checker *compiler.Checker, sourceFile *compiler.Node) []SemanticTokenEntry {
	text := sourceFile.AsSourceFile().Text
	var entries []SemanticTokenEntry
	var visit compiler.Visitor
	visit = func(node *compiler.Node) bool {
		if node.Kind == compiler.Identifier {
			symbol := checker.ResolveName(node)
			tokenType := -1
			if symbol != nil {
				tokenType = tokenTypeOf(symbol.Flags)
			}
			if symbol != nil && tokenType != -1 {
				start := compiler.SkipTrivia(text, node.Pos)
				length := node.End - start
				if length > 0 {
					isDeclarationName := node.Parent != nil && node.Parent.Symbol == symbol &&
						nodeNameOf(node.Parent) == node
					entries = append(entries, SemanticTokenEntry{
						Offset:         start,
						Length:         length,
						TokenType:      tokenType,
						TokenModifiers: modifiersOf(symbol, isDeclarationName),
					})
				}
			}
		}
		if node.Kind == compiler.CallExpression {
			if method := checker.ResolveCall(node); method != nil {
				methodName := ""
				if mn := declName(method); mn != nil {
					methodName = mn.AsIdentifier().Text
				}
				if indices, ok := regexSinks[enclosingTypeName(method)+"#"+methodName]; ok {
					if args := node.AsCallExpression().Arguments; args != nil {
						for i, arg := range args.Nodes {
							if indices[i] && arg.Kind == compiler.StringLiteral {
								start := compiler.SkipTrivia(text, arg.Pos)
								if length := arg.End - start; length > 0 {
									entries = append(entries, SemanticTokenEntry{Offset: start, Length: length, TokenType: typeIndex("regexp"), TokenModifiers: 0})
								}
							}
						}
					}
				}
			}
		}
		node.ForEachChild(visit)
		return false
	}
	visit(sourceFile)
	slices.SortStableFunc(entries, func(a, b SemanticTokenEntry) int { return cmp.Compare(a.Offset, b.Offset) })
	return entries
}

// nodeNameOf returns the name Identifier of a node (for declaration-name checks).
func nodeNameOf(node *compiler.Node) *compiler.Node {
	return declName(node)
}

File: src/services/semanticTokens.ts
// Semantic tokens: classify every resolved identifier so editors color fields,
// locals, parameters, type parameters etc. accurately. Offset-based and
// position-free; the LSP server encodes the entries into the wire format.

import type { Checker } from "../compiler/checker.ts";
import { symbolDeprecation } from "../compiler/deprecation.ts";
import { forEachChild } from "../compiler/parser.ts";
import { getSourceFileOfNode } from "../compiler/resolver.ts";
import {
  type CallExpression,
  type Identifier,
  type Node,
  type SourceFile,
  type Symbol,
  SymbolFlags,
  SyntaxKind,
} from "../compiler/types.ts";
import { skipTrivia } from "../compiler/utilities.ts";
import { isSyntheticUri } from "../workspace.ts";

// The legend the server advertises; indexes below refer into these arrays.
export const TOKEN_TYPES = [
  "namespace",
  "class",
  "interface",
  "enum",
  "enumMember",
  "typeParameter",
  "type",
  "method",
  "property",
  "parameter",
  "variable",
  "regexp",
] as const;

export const TOKEN_MODIFIERS = [
  "declaration",
  "static",
  "readonly",
  "defaultLibrary",
  "deprecated",
] as const;

const TYPE_INDEX = Object.fromEntries(TOKEN_TYPES.map((t, i) => [t, i]));
const MOD_BIT = Object.fromEntries(TOKEN_MODIFIERS.map((m, i) => [m, 1 << i]));

export interface SemanticTokenEntry {
  readonly offset: number;
  readonly length: number;
  readonly tokenType: number; // index into TOKEN_TYPES
  readonly tokenModifiers: number; // bit set over TOKEN_MODIFIERS
}

function tokenTypeOf(flags: SymbolFlags): number | undefined {
  if (flags & SymbolFlags.Package) return TYPE_INDEX["namespace"];
  if (flags & SymbolFlags.Class) return TYPE_INDEX["class"];
  if (flags & SymbolFlags.Interface) return TYPE_INDEX["interface"];
  if (flags & SymbolFlags.Enum) return TYPE_INDEX["enum"];
  if (flags & SymbolFlags.Record) return TYPE_INDEX["class"];
  if (flags & SymbolFlags.Annotation) return TYPE_INDEX["type"];
  if (flags & SymbolFlags.TypeParameter) return TYPE_INDEX["typeParameter"];
  if (flags & SymbolFlags.EnumConstant) return TYPE_INDEX["enumMember"];
  if (flags & (SymbolFlags.Method | SymbolFlags.Constructor)) return TYPE_INDEX["method"];
  if (flags & SymbolFlags.Field) return TYPE_INDEX["property"];
  if (flags & SymbolFlags.Parameter) return TYPE_INDEX["parameter"];
  if (flags & SymbolFlags.LocalVariable) return TYPE_INDEX["variable"];
  return undefined;
}

// The node whose modifiers govern this symbol: the field declaration for a
// declarator, otherwise the declaration itself.
function modifierCarrier(symbol: Symbol): { modifiers?: readonly Node[] } | undefined {
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
  if (!declaration) return undefined;
  if (declaration.kind === SyntaxKind.VariableDeclarator) {
    return declaration.parent as { modifiers?: readonly Node[] };
  }
  return declaration as { modifiers?: readonly Node[] };
}

function modifiersOf(symbol: Symbol, isDeclarationName: boolean): number {
  let bits = isDeclarationName ? MOD_BIT["declaration"]! : 0;
  if (symbol.flags & SymbolFlags.EnumConstant) {
    bits |= MOD_BIT["static"]! | MOD_BIT["readonly"]!;
  } else {
    const carrier = modifierCarrier(symbol);
    for (const m of carrier?.modifiers ?? []) {
      if (m.kind === SyntaxKind.StaticKeyword) bits |= MOD_BIT["static"]!;
      if (m.kind === SyntaxKind.FinalKeyword) bits |= MOD_BIT["readonly"]!;
    }
  }
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
  if (declaration && isSyntheticUri(getSourceFileOfNode(declaration).fileName)) {
    bits |= MOD_BIT["defaultLibrary"]!;
  }
  if (symbolDeprecation(symbol)) bits |= MOD_BIT["deprecated"]!;
  return bits;
}

// Standard-library methods whose named string parameters are regular
// expressions, keyed by "OwnerSimpleName#method" -> regex argument indices.
// A string literal passed at one of these positions is tokenized as `regexp`
// so clients can highlight the pattern (Java has no regex literal syntax).
const REGEX_SINKS: Record<string, ReadonlySet<number>> = {
  "Pattern#compile": new Set([0]),
  "Pattern#matches": new Set([0]),
  "String#matches": new Set([0]),
  "String#split": new Set([0]),
  "String#replaceAll": new Set([0]),
  "String#replaceFirst": new Set([0]),
};

// The simple name of the type declaration enclosing a declaration node.
function enclosingTypeName(node: Node): string | undefined {
  for (let n = node.parent; n; n = n.parent) {
    switch (n.kind) {
      case SyntaxKind.ClassDeclaration:
      case SyntaxKind.InterfaceDeclaration:
      case SyntaxKind.EnumDeclaration:
      case SyntaxKind.RecordDeclaration:
      case SyntaxKind.AnnotationTypeDeclaration:
        return (n as { name?: Identifier }).name?.text;
    }
  }
  return undefined;
}

export function getSemanticTokens(checker: Checker, sourceFile: SourceFile): SemanticTokenEntry[] {
  const entries: SemanticTokenEntry[] = [];
  const visit = (node: Node): void => {
    if (node.kind === SyntaxKind.Identifier) {
      const symbol = checker.resolveName(node as Identifier);
      const tokenType = symbol ? tokenTypeOf(symbol.flags) : undefined;
      if (symbol && tokenType !== undefined) {
        const start = skipTrivia(sourceFile.text, node.pos);
        const length = node.end - start;
        if (length > 0) {
          const isDeclarationName =
            !!node.parent &&
            node.parent.symbol === symbol &&
            (node.parent as { name?: Node }).name === node;
          entries.push({
            offset: start,
            length,
            tokenType,
            tokenModifiers: modifiersOf(symbol, isDeclarationName),
          });
        }
      }
    }
    if (node.kind === SyntaxKind.CallExpression) {
      const call = node as CallExpression;
      const method = checker.resolveCall(call);
      const indices = method && REGEX_SINKS[`${enclosingTypeName(method)}#${method.name?.text}`];
      if (indices) {
        for (const i of indices) {
          const arg = call.arguments[i];
          if (arg?.kind === SyntaxKind.StringLiteral) {
            const start = skipTrivia(sourceFile.text, arg.pos);
            const length = arg.end - start;
            if (length > 0) {
              entries.push({
                offset: start,
                length,
                tokenType: TYPE_INDEX["regexp"]!,
                tokenModifiers: 0,
              });
            }
          }
        }
      }
    }
    forEachChild(node, child => {
      visit(child);
      return undefined;
    });
  };
  visit(sourceFile);
  entries.sort((a, b) => a.offset - b.offset);
  return entries;
}
